### ISSUE TITLE
Objection 2.x does not support knex 0.95

### ISSUE THREAD

**Issue by slim-hmidi** [knex/knex#4389]
# Environment

Knex version: 0.95.2
Database + version: postgresql 13
OS: windows 10
typescript: 4.2.3
ts-node: 9.1.1
nodemon: 2.0.7

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do
When I launch the app with nodemon I got many errors concerning knex/types
2. Error message
```
[start:build] node_modules/objection/typings/objection/index.d.ts(122,32): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'Raw'.
[start:build] node_modules/objection/typings/objection/index.d.ts(527,49): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'JoinClause'.
[start:build] node_modules/objection/typings/objection/index.d.ts(613,22): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'ColumnInfo'.
[start:build] node_modules/objection/typings/objection/index.d.ts(673,32): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1071,45): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1187,34): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'Transaction'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1188,49): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1382,12): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1459,17): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1459,24): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1460,23): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1596,24): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1596,31): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1597,30): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1740,14): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1741,21): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1755,24): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1833,13): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1839,12): error TS2709: Cannot use namespace 'Knex' as a type.
```

**Comment by kibertoad** [comment]
Please check the migration guide: https://github.com/knex/knex/blob/master/UPGRADING.md
There were breaking TS changes.

**Comment by kibertoad** [comment]
Note that Objection will only support knex 0.95+ in 3.0 which is yet to be released.

**Comment by kibertoad** [comment]
We are waiting for https://github.com/Vincit/objection.js/pull/1992 to be merged.

**Comment by slim-hmidi** [comment]
@kibertoad thanks for your answer.

**Comment by TheMeteoRain** [comment]
Not sure if I'm barking the same tree, so I won't make a new issue.

The namespace that knex is providing to import is kinda odd. I have checked multiple times that it should work and that it's exported correctly and pointing to the right namespace. Yet I keep getting this error.

```
$ yarn knex --debug migrate:make migration_create_table -x ts 
yarn run v1.22.10
Requiring external module ts-node/register

/Users/akash.singh/Projects/Personal/mete/node_modules/ts-node/src/index.ts:513
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
knexfile.ts:5:71 - error TS2694: Namespace '"node_modules/knex/types/index".Knex' has no exported member 'Config'.

5 export type KnexEnvironmentConfig = Record<DatabaseEnvironments, Knex.Config>
```

This is how I import my Knex namespace

`import { Knex } from 'knex'
`

**Comment by kibertoad** [comment]
Which TypeScript version are you using?

**Comment by kibertoad** [comment]
@TheMeteoRain See the migration guide: https://github.com/knex/knex/blob/master/UPGRADING.md
4.1+ is needed.

**Comment by TheMeteoRain** [comment]
Really!? I did read that just before commenting, but seems like I missed that part.
First time i've come a cross typescript version requirement hitting so close to my own.

I've just updated my typescript. Seems to be working now. Well, thanks for that!

**Comment by ps2goat** [comment]
were 0.94.x versions removed?!

**Comment by HamzaKV** [comment]
Facing same issue.

Environment:
- typescript: 4.4.3
- objection: 2.2.16
- pg: 8.7.1
- knex: 0.95.11
- Node: 16.8.0

```
[start:build] node_modules/objection/typings/objection/index.d.ts(122,32): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'Raw'.
[start:build] node_modules/objection/typings/objection/index.d.ts(527,49): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'JoinClause'.
[start:build] node_modules/objection/typings/objection/index.d.ts(613,22): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'ColumnInfo'.
[start:build] node_modules/objection/typings/objection/index.d.ts(673,32): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1071,45): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1187,34): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'Transaction'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1188,49): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1382,12): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1459,17): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1459,24): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1460,23): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1596,24): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1596,31): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1597,30): error TS2694: Namespace '"C:/projects/myProject/backend/node_modules/knex/types/index"' has no exported member 'QueryBuilder'.
[start:build] node_modules/objection/typings/objection/index.d.ts(1740,14): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1741,21): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1755,24): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1833,13): error TS2709: Cannot use namespace 'Knex' as a type.
[start:build] node_modules/objection/typings/objection/index.d.ts(1839,12): error TS2709: Cannot use namespace 'Knex' as a type.

```
Following the guide did not help.
Solved it by downgrading knex to less than 0.95.0

**Comment by OlivierCavadenti** [comment]
Objection 3.0.0 is now out https://github.com/Vincit/objection.js/releases/tag/3.0.0
And support Knex 0.95.13 at this time: 
https://github.com/Vincit/objection.js/blob/f5ad7b9b2739b8e958e20680338fcdd9a8e7ed34/package.json#L70
